### PR TITLE
Next skip

### PR DESCRIPTION
**Pull request by samuelcolvin**
Support skipping vales with `Jiter`, this is specifically for https://github.com/apache/datafusion/issues/7845.

The performance seems good, benchmarks show skipping values is significantly faster than not.

Running the query from https://github.com/apache/datafusion/issues/7845#issuecomment-2061121332, gives:

```sql
-- datafusion with next_value
SELECT count(*) FROM records where json_contains(attributes, 'size');
  -- 6165586 in 17.2s

-- datafusion with next_skip
SELECT count(*) FROM records where json_contains(attributes, 'size');
  -- 6165586 in 14.3s

-- datafusion using serde-json allocating keys to a Vec<String> then checking for the key
SELECT count(*) FROM records where json_contains(attributes, 'size');
  -- 6165586 in 21.7s

-- datafusion return false - e.g. fastest case for a UDF
SELECT count(*) FROM records where json_contains(attributes, 'size');
  -- 6165586 in 11.8s

-- datafusion like
SELECT count(*) FROM records where attributes like '%"size":%';
  -- 6165586 in 14.5s

-- duckdb
SELECT count(*) FROM read_parquet('file.parquet') where list_contains(json_keys(attributes), 'size')
  -- 6165586 in 14.2s
```

### CODE REVIEW

File: crates/jiter/src/number_decoder.rs [hunk 1]
```diff
@@ -384,8 +384,8 @@ impl AbstractNumberDecoder for NumberRange {
                         let end = consume_exponential(data, index)?;
                         Ok((start..end, end))
                     }
-                    Some(_) => return json_err!(InvalidNumber, index),
-                    None => return Ok((start..index, index)),
+                    Some(digit) if digit.is_ascii_digit() => json_err!(InvalidNumber, index),
```
Comment (davidhewitt) [inline crates/jiter/src/number_decoder.rs:387]: Doesn't this mean that things like `0a` are now unintentionally valid?

File: crates/jiter/src/parse.rs [hunk 2]
```diff
@@ -19,7 +19,7 @@ impl Peek {
 }
 
 impl Peek {
-    const fn new(next: u8) -> Self {
+    pub const fn new(next: u8) -> Self {
```
Comment (davidhewitt) [inline crates/jiter/src/parse.rs:22]: ?

```suggestion
    pub(crate) const fn new(next: u8) -> Self {
```

File: crates/jiter/src/number_decoder.rs [hunk 3]
```diff
@@ -398,25 +398,49 @@ impl AbstractNumberDecoder for NumberRange {
...
                     let end = consume_exponential(data, index)?;
                     return Ok((start..end, end));
                 }
-                _ => break,
             }
-            index += 1;
+            return Ok((start..index, index));
+        }
+        loop {
+            let (chunk, new_index) = IntChunk::parse_big(data, index);
```
Comment (davidhewitt) [inline crates/jiter/src/number_decoder.rs:419]: As I understand it, this is just adding the SIMD code to range to allow skipping to be faster?

File: crates/jiter/src/number_decoder.rs (hunk 1 above)
Comment (samuelcolvin) [reply crates/jiter/src/number_decoder.rs:387]: no, the next step takes care, this matches the normal number parser.

File: crates/jiter/src/parse.rs (hunk 2 above)
Comment (samuelcolvin) [reply crates/jiter/src/parse.rs:22]: no, needed for tests.

File: crates/jiter/src/number_decoder.rs (hunk 3 above)
Comment (samuelcolvin) [reply crates/jiter/src/number_decoder.rs:419]: yes.

**Pull request merged**